Sources/SourceKitLSP/CapabilityRegistry.swift

```diff
@@ -0,0 +1,90 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import LanguageServerProtocol
+
+/// A class which tracks the client's capabilities as well as our dynamic
+/// capability registrations in order to avoid registering conflicting
+/// capabilities.
+public final class CapabilityRegistry {
+  /// Registered completion options.
+  private var completion: [Registration: CompletionRegistrationOptions] = [:]
+
+  /// Whether dynamic registration is supported for completion.
+  public var completionSupported: Bool = false
+
+  private(set) var clientCapabilities: ClientCapabilities
+
+  public init(clientCapabilities: ClientCapabilities) {
+    self.clientCapabilities = clientCapabilities
+  }
+
+  public func registerCompletion(
```

Comment:
We should clarify the contract here with respect to whether dynamic registration is supported.  Naively I would expect either `precondition(completionSupported)` or `guard completionSupported else { return nil }` or something like that.

Reply:
Now this is done inside the Registry